### PR TITLE
[serve] Enable streaming over `ServeHandle`s

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Adds support for `handle.options(stream=True)` to make streaming calls using the `RayServeHandle`.

This required two changes:

- Adding `stream=True` to the `.options` codepaths.
- Adding a branch in `handle_request_streaming` to handle calls from the HTTP proxy as well as other handle calls.

I also refactored some of the code around exception handling/metrics into a common context manager that's used by both codepaths.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/handle.py [hunk 1]
```diff
@@ -388,17 +397,34 @@ def __init__(
...
-    def options(self, *, method_name: str) -> "RayServeDeploymentHandle":
-        return self.__class__(
-            self.deployment_name, HandleOptions(method_name=method_name)
-        )
+    def options(
+        self,
+        *,
+        method_name: Union[str, DEFAULT] = DEFAULT.VALUE,
+        stream: Union[bool, DEFAULT] = DEFAULT.VALUE,
+    ) -> "RayServeDeploymentHandle":
```
Comment (edoakes) [inline python/ray/serve/handle.py:405]: TODO: make all of the `.options` go through the same codepath

**Comment by edoakes** [comment]
https://github.com/ray-project/ray/issues/37147 causes the `aiowrlock` reader lock to raise an exception. It's currently commented out; once the fix for that issue is in, I'll uncomment and this can be merged.

File: python/ray/serve/_private/replica.py [hunk 2]
```diff
@@ -260,24 +262,19 @@ async def handle_request_streaming(
             interface. This allows us to return the messages back to the HTTP proxy as
             they're sent by user code (e.g., the FastAPI wrapper).
             """
-            request_metadata = pickle.loads(pickled_request_metadata)
-            if not request_metadata.is_http_request:
-                raise NotImplementedError(
-                    "Only HTTP requests are currently supported over streaming."
-                )
-
+            # TODO: update comment ^^^
```
Comment (GeneDer) [inline python/ray/serve/_private/replica.py:265]: Should this be updated together? 🙃

File: python/ray/serve/_private/replica.py [hunk 3]
```diff
@@ -692,44 +641,140 @@ async def update_user_config(self, user_config: Any):
...
+        )
+
+        logger.info(
+            f"Started executing request {request_metadata.request_id}",
+            extra={"log_to_stderr": False},
+        )
+        start_time = time.time()
+        user_exception = None
+        try:
+            # TODO: WTF is up with this thing.
```
Comment (GeneDer) [inline python/ray/serve/_private/replica.py:673]: Does this still applies? lol

File: python/ray/serve/_private/replica.py [hunk 4]
```diff
@@ -692,44 +641,140 @@ async def update_user_config(self, user_config: Any):
...
+            # with self.rwlock.reader:
+            yield
+        except Exception as e:
+            user_exception = e
+            logger.exception(f"Request failed due to {type(e).__name__}:")
+            if ray.util.pdb._is_ray_debugger_enabled():
+                ray.util.pdb._post_mortem()
```
Comment (GeneDer) [inline python/ray/serve/_private/replica.py:680]: Just out of curiosity, what does this do?

File: python/ray/serve/tests/test_metrics.py [hunk 5]
```diff
@@ -893,6 +893,7 @@ def metric_available() -> bool:
             metric_dict_str = f"dict({line[dict_body_start:dict_body_end]})"
             metric_dicts.append(eval(metric_dict_str))
 
+    print(metric_dicts)
```
Comment (GeneDer) [inline python/ray/serve/tests/test_metrics.py:896]: Can probably delete this 🙃

File: python/ray/serve/tests/test_util.py [hunk 6]
```diff
@@ -618,6 +619,41 @@ def test_calculate_remaining_timeout():
     )
 
 
+@pytest.mark.asyncio
+async def test_wrap_generator_function_in_async_if_needed():
```
Comment (GeneDer) [inline python/ray/serve/tests/test_util.py:623]: TIL: tests can be async too 😯

**Review by GeneDer — APPROVED** [review]
LGTM!

File: python/ray/serve/_private/replica.py (hunk 4 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:680]: It is so people can use the ray debugger: https://docs.ray.io/en/latest/ray-observability/user-guides/debug-apps/ray-debugging.html#ray-debugger

To be honest I don't think anyone is using this with Serve. The feature is pretty half baked from what I understand.

File: python/ray/serve/_private/replica.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:673]: lol it does at the moment, it's related to the bug I linked above

File: python/ray/serve/_private/replica.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:265]: yep let me clean this up

File: python/ray/serve/_private/replica.py (hunk 4 above)
Comment (GeneDer) [reply python/ray/serve/_private/replica.py:680]: oh cool this is good to know!

File: python/ray/serve/handle.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/handle.py:405]: will do this in a separate PR

Comment (edoakes) [reply python/ray/serve/handle.py:405]: https://github.com/ray-project/ray/pull/37199

**Pull request merged**